core/api_contacts.go

```diff
@@ -0,0 +1,90 @@
+package core
+
+import (
+	"net/http"
+
+	"github.com/gin-gonic/gin"
+)
+
+func (a *api) lsContacts(g *gin.Context) {
+	opts, err := a.readOpts(g)
+	if err != nil {
+		a.abort500(g, err)
+		return
+	}
+
+	var contacts []ContactInfo
+
+	threadId := opts["thread"]
+	if threadId == "default" {
+		threadId = a.node.config.Threads.Defaults.ID
+	}
+	if threadId != "" {
+		thrd := a.node.Thread(threadId)
+		if thrd == nil {
+			g.String(http.StatusNotFound, ErrThreadNotFound.Error())
+			return
+		}
+		contacts = make([]ContactInfo, 0)
+		for _, p := range thrd.Peers() {
+			contact := a.node.Contact(p.Id)
+			if contact != nil {
+				contacts = append(contacts, *contact)
+			}
+		}
+	} else {
+		contacts, err = a.node.Contacts()
+	}
+
+	g.JSON(http.StatusOK, contacts)
+}
+
+func (a *api) getContacts(g *gin.Context) {
+	id := g.Param("id")
+
+	info := a.node.Contact(id)
+	if info == nil {
+		g.String(http.StatusNotFound, "contact not found")
+		return
+	}
+
+	g.JSON(http.StatusOK, info)
+}
+
+func (a *api) addContacts(g *gin.Context) {
+	args, err := a.readArgs(g)
+	if err != nil {
+		a.abort500(g, err)
+		return
+	}
+	if len(args) < 2 {
+		g.String(http.StatusBadRequest, "missing peer id or address")
+		return
+	}
+	opts, err := a.readOpts(g)
+	if err != nil {
+		a.abort500(g, err)
+		return
+	}
+
+	id := args[0]
+	username := opts["username"]
+	if username == "" {
+		username = id[len(id)-7:]
```

Comment:
wouldn't this be done _on the way out_? as in, when getting the contact, if username == "", use peerId